runtime/src/accounts_db.rs

```diff
@@ -2150,27 +2762,231 @@ impl AccountsDB {
+        }
+
+        // Only add to the uncleaned roots set *after* we've flushed the previous roots,
+        // so that clean will actually be able to clean the slots.
+        self.accounts_index.add_uncleaned_roots(cached_roots);
+        flush_roots_elapsed.stop();
+
+        // Note we don't purge unrooted slots here because there may be ongoing scans/references
+        // for those slot, let the Bank::drop() implementation do cleanup instead on dead
+        // banks
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        let old_slots = self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS);
+        let total_excess_slot_count = old_slots.len();
+        let mut unflushable_unrooted_slot_count = 0;
+        let max_flushed_root = self.accounts_cache.fetch_max_flush_root();
+        for old_slot in old_slots {
+            // Don't flush slots that are known to be unrooted
+            if old_slot > max_flushed_root {
+                self.flush_slot_cache(old_slot);
+            } else {
+                unflushable_unrooted_slot_count += 1;
+            }
+        }
+
+        datapoint_info!(
+            "accounts_db-cache-limit-slots",
+            ("total_excess_slot_count", total_excess_slot_count, i64),
+            (
+                "unflushable_unrooted_slot_count",
+                unflushable_unrooted_slot_count,
+                i64
+            ),
+        );
+
+        // Flush a random slot out after every force flush to catch any inconsistencies
+        // between cache and written state (i.e. should cause a hash mismatch between validators
+        // that flush and don't flush if such a bug exists).
+        let num_slots_remaining = self.accounts_cache.num_slots();
+        if force_flush && num_slots_remaining >= FLUSH_CACHE_RANDOM_THRESHOLD {
+            // Don't flush slots that are known to be unrooted
+            let mut frozen_slots = self.accounts_cache.find_older_frozen_slots(0);
+            frozen_slots.retain(|s| *s > max_flushed_root);
+            // Remove a random index 0 <= i < `frozen_slots.len()`
+            let rand_slot = frozen_slots[thread_rng().gen_range(0, frozen_slots.len())];
+            info!(
+                "Flushing random slot: {}, num_remaining: {}",
+                rand_slot, num_slots_remaining
+            );
+            self.flush_slot_cache(rand_slot);
+        }
+
+        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    }
+
+    fn flush_slot_cache(&self, slot: Slot) {
+        info!("flush_slot_cache slot: {}", slot);
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            let iter_items: Vec<_> = slot_cache.iter().collect();
+            let mut total_size = 0;
+            let (accounts, hashes): (Vec<(&Pubkey, &Account)>, Vec<Hash>) = iter_items
+                .iter()
+                .map(|iter_item| {
```

Comment:
considering this size concern https://github.com/solana-labs/solana/pull/13140/files#r553473450, now that we're better off doing shrinking here in this pr altogether as well?.

Yeah, I dragged you keeping this pr open for too long... But, requiring new hunk of heap 34G and whooping write of 34G to fs (could be tmpfs)  as-is  would be just another op nightmare.

Reply:
now that we packed the flushing timing shift, introduced randomness, added runtime flag knob.

I think we have plenty reasons to justify to forge shipping this not across several point releases. :)

Reply:
I'm relaxing this now a bit, now that I remembered this pr as-is should reduce intra-slot updates, resulting 50% IO writes..